### PR TITLE
backport a few bugfixes from master

### PR DESCRIPTION
**Pull request by bergwolf**
The PR backports following commits from master:

4d0f3c5f vc: make host shared path readonly

### CODE REVIEW

File: virtcontainers/mount.go [hunk 1]
```diff
@@ -328,12 +343,25 @@ type Mount struct {
...
+
+func bindUnmountContainerRootfs(ctx context.Context, sharedDir, cID string) error {
 	span, _ := trace(ctx, "bindUnmountContainerRootfs")
 	defer span.Finish()
 
-	rootfsDest := filepath.Join(sharedDir, sandboxID, cID, rootfsDir)
-	err := syscall.Unmount(rootfsDest, syscall.MNT_DETACH)
+	rootfsDest := filepath.Join(sharedDir, cID, rootfsDir)
+	if isSymlink(filepath.Join(sharedDir, cID)) || isSymlink(rootfsDest) {
+		logrus.Warnf("container dir %s is a symlink, malicious guest?", cID)
```
Comment (tedyu) [inline virtcontainers/mount.go:360]: cID itself is not directory.
It seems this sentence can be made clearer.

**Comment by tedyu** [comment]
lgtm

**Comment by amshinde** [comment]
Restating failing CIs.

@jcvenegas Which of the CLH jobs are expected to fail?

**Comment by amshinde** [comment]
@bergwolf @chavafg Seeing this failure for the crio test on ubuntu: 
```
16:54:12 not ok 3 ctr termination reason Error
16:54:12 # (from function `setup_crio' in file helpers.bash, line 261,
16:54:12 #  from function `start_crio' in file helpers.bash, line 331,
16:54:12 #  in test file ctr.bats, line 46)
16:54:12 #   `start_crio' failed
16:54:12 # time="2020-06-03T23:54:08Z" level=error msg="error opening storage: /dev/sdb is already part of a volume group \"storage\": must remove this device from any volume group or provide a different device"
16:54:12 # time="2020-06-03T23:54:10Z" level=fatal msg="failed to connect: failed to connect, make sure you are running as root and the runtime has been started: context deadline exceeded"
16:54:12 # time="2020-06-03T23:54:12Z" level=fatal msg="failed to connect: failed to connect, make sure you are running as root and the runtime has been started: context deadline exceeded"
16:54:12 # 0
```
I'll restart this one more time, if it fails could be a legitimate that needs investigating.

**Comment by bergwolf** [comment]
Looks like some test setup failure. How are we testing stable branch PRs. Are we using corresponding stable branch tests?

```
./rootfs-builder/rootfs.sh: line 280: /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-PR-initrd/go/src/github.com/kata-containers/osbuilder/rootfs-builder/*default-initrd-name/config.sh: No such file or directory
Failed at 118: sudo -E AGENT_INIT="${AGENT_INIT}" AGENT_VERSION="${agent_commit}" GOPATH="$GOPATH" USE_DOCKER=true OS_VERSION=${os_version} ./rootfs-builder/rootfs.sh "${distro}"
Failed at 20: "${cidir}/install_kata_image.sh" "${tag}"
Build step 'Execute shell' marked build as failure
Performing Post build task...
Match found for :.* : True
Logical operation result is TRUE
Running script  : #!/bin/bash
```

**Comment by chavafg** [comment]
The cri-o issue only happened on ubuntu as we use block devices there, but I do not understand why it failed. The teardown cleans the volume group after each tests, so we shouldn't have that issue. On master it seems that it is working correctly. Will try to debug locally.

**Comment by chavafg** [comment]
@bergwolf, 

I have tested locally with stable-1.10 branch (without this PR) and works correctly. I bisected commits included here and the one that make this fail is 20759e178144998506be6b143443ffb376e461b5. Without this commit tests pass correctly, maybe something else is missing to make this commit work correctly.

**Comment by amshinde** [comment]
@bergwolf I have opened https://github.com/kata-containers/runtime/pull/2744 to include the other commits to see if CI passes.
@chavafg mentioned looks like https://github.com/kata-containers/runtime/commit/20759e178144998506be6b143443ffb376e461b5 is the cause of failure. 
Maybe the way the directories are setup is different in 1.10. Needs to be looked into.

File: virtcontainers/mount.go (hunk 1 above)
Comment (bergwolf) [reply virtcontainers/mount.go:360]: It is a directory name.

Comment (jodh-intel) [reply virtcontainers/mount.go:360]: @tedyu - just to be clear - this is a backport PR: these changes have already landed in `master` so  we want to keep the backport changes as similar to the original changes to `master` as possible.

**Comment by jodh-intel** [comment]
Restarted the Ubuntu CI which failed in an interesting way!:

```
not ok 3 ctr termination reason Error
18:03:22 # (from function `setup_crio' in file helpers.bash, line 261,
18:03:22 #  from function `start_crio' in file helpers.bash, line 331,
18:03:22 #  in test file ctr.bats, line 46)
18:03:22 #   `start_crio' failed
18:03:22 # time="2020-06-06T17:03:18Z" level=error msg="error opening storage: /dev/sdb is already part of a volume group \"storage\": must remove this device from any volume group or provide a different device"
```

**Comment by amshinde** [comment]
@bergwolf I took a look at this to reproduce and look into this issue a bit.
What I am seeing is that the failure is essentially caused by the call to function `bindUnmountContainerRootfs`. If I uncomment this function, `make crio` passes. 

In case of devicemapper, function `bindContainerRootfs` is never called, we mount the block device inside the guest using storage api. So I would assume that a call to `bindUnmountContainerRootfs` would not really unmount any bind-mounts. I havent looked at your code too closely, but apparently this function does perform unmounts that mess up devicemapper. 
Do you think you can take a look at this?  I can continue debugging a bit tomorrow, but I do want to tag release on 1.10 branch tomorrow PST.

**Comment by jcvenegas** [comment]
@amshinde I could not see the logs for clh, anyway I added backports for a kernel patch that brings more stability to kata + clh. 
https://github.com/kata-containers/packaging/pull/1074
https://github.com/kata-containers/packaging/pull/1073

**Comment by bergwolf** [comment]
@amshinde @jcvenegas Thanks for your help. It should be fixed now.  Right now bindUnmountContainerRootfs() is unconditionally called and dm based rootfs should just ignore the errors.

**Pull request merged**